File: leetcode/medium/1011/1011.ts
function shipWithinDays(weights: number[], days: number): number {
  let left = Math.max(...weights);
  let right = weights.reduce((sum, w) => sum + w, 0);
  let result = right;

  function canShip(capacity: number): boolean {
    let ships = 1;
    let currentCapacity = capacity;

    for (const weight of weights) {
      if (currentCapacity - weight < 0) {
        ships++;
        currentCapacity = capacity;
      }
      currentCapacity -= weight;
    }

    return ships <= days;
  }

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);

    if (canShip(mid)) {
      result = Math.min(result, mid);
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return result;
}

File: leetcode/medium/1011/1011.rs
impl Solution {
    pub fn ship_within_days(weights: Vec<i32>, days: i32) -> i32 {
        let mut left = *weights.iter().max().unwrap();
        let mut right: i32 = weights.iter().sum();
        let mut result = right;

        let can_ship = |capacity: i32| -> bool {
            let mut ships = 1;
            let mut current_capacity = capacity;

            for &weight in &weights {
                if current_capacity - weight < 0 {
                    ships += 1;
                    current_capacity = capacity;
                }
                current_capacity -= weight;
            }

            ships <= days
        };

        while left <= right {
            let mid = (left + right) / 2;

            if can_ship(mid) {
                result = result.min(mid);
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        result
    }
    
}